### PR TITLE
[GH-709] Remove reference to Chain.chain_state() in ContractCallTx

### LINKED ISSUE #709: Remove reference to Chain.chain_state() in contract call tx

### PR DESCRIPTION
**Pull request by d-velev**
closes #709

### CODE REVIEW

File: apps/aecore/lib/aecore/tx/data_tx.ex [hunk 1]
```diff
@@ -288,7 +288,14 @@ defmodule Aecore.Tx.DataTx do
         context \\ :transaction
       ) do
     accounts_state = chainstate.accounts
-    tx_type_state = Map.get(chainstate, type.get_chain_state_name(), %{})
+    chain_state_name = type.get_chain_state_name()
+
+    tx_type_state =
+      if Enum.member?([:contracts, :calls], chain_state_name) do
+        chainstate
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/tx/data_tx.ex]: you require the full chain state for this, right?

File: apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex [hunk 2]
```diff
@@ -217,16 +216,14 @@ defmodule Aecore.Contract.Tx.ContractCallTx do
         ) :: :ok | {:error, String.t()}
   def preprocess_check(
         accounts,
-        _calls,
+        chainstate,
```
Comment (DanielaIvanova) [inline apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex]: The second argument supposed to be "Structure that holds specific transaction info in the chainstate". I am not sure if this is correct. You can check [here](https://github.com/aeternity/elixir-node/blob/master/apps/aecore/lib/aecore/tx/transaction.ex#L51).

File: apps/aecore/lib/aecore/tx/data_tx.ex (hunk 1 above)
Comment (thepiwo) [reply apps/aecore/lib/aecore/tx/data_tx.ex]: :+1:

File: apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex (hunk 2 above)
Comment (thepiwo) [reply apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex]: Yes @DanielaIvanova is right, the `@spec` of the behaviour is broken with this

Comment (d-velev) [reply apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex]: yes I know but we need the entire chainstate for this one, not only calls

Comment (thepiwo) [reply apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex]: yes, but as you do it won't work with the spec

Comment (d-velev) [reply apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex]: right

**Comment by d-velev** [comment]
@DanielaIvanova The reason is that we don't want to take the current chainstate from Chain.chain_state, we want to work with the copy of the chainstate from when we know it was the correct one

File: apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex [hunk 3]
```diff
@@ -208,7 +208,7 @@ defmodule Aecore.Contract.Tx.ContractCallTx do
 
   @spec preprocess_check(
           Chainstate.accounts(),
-          tx_type_state(),
+          Chainstate.t(),
```
Comment (thepiwo) [inline apps/aecore/lib/aecore/contract/tx/contract_call_tx.ex]: this is still not the correct spec, as the behaviour callbacks state `tx_type_state()`

**Pull request merged**